docs/architecture/017-applicant-ux.md

```diff
@@ -0,0 +1,89 @@
+# 017: Applicant Form UX
+
+Authors: @opqdonut, @Macroz, @marharyta
+
+This ADR aims to collect together some topics like validations in
+frontend, autosaving, etc. that are related to the Applicant's user
+experience. There have been previous discussions about all of these
+topics but no unified presentation.
+
+Based on ticket [#2639](https://github.com/CSCfi/rems/issues/2639)
+
+# Problems with the current applicant UX
+
+## A) Validation errors block saving
+
+You don't know if your inputs (e.g. ip address) are valid before
+pressing Save, and then you need to fix all of these errors before
+REMS saves your work.
+
+This can be irritating if you're working on a long application, and
+just want to save your work and go home. It's less of an issue if
+users save early & often.
+
+**Possible solution**: Show validation errors to the user, but perform
+the save anyway. Show errors in yellow?
+
+Historical note: historically (in the pre-SPA days) REMS used to work
+like this. This topic was more recently touched on in
+[#2117](https://github.com/CSCfi/rems/issues/2117). Interestingly
+enough, the idea idea seems to have been to only perform schema-style
+validations on drafts. However, the email content validation was
+included in the spec, so subsequent field types added their content
+validations to save-draft (see e.g.
+[#2567](https://github.com/CSCfi/rems/pull/2567),
+[#2594](https://github.com/CSCfi/rems/pull/2594))
+
+Idea: Think of REMS validations more like spellchecking and less like
+validations. Higlight in red things that the applicant will want to
+fix, but don't prevent saving.
+
+## B) Validation errors are shown late
+
+You need to perform an action (save/submit) to see the validation
+errors. Technically we should be able to run the validation code on
+the frontend at any point. There's no need to talk to the backend.
+
+**Possible solution**: Validate when user stops typing / leaves field.
+
+Questions:
+- What about partially filled in forms? It would be best to not show validation errors for fields that haven't been edited yet.
+- What if a user navigates away from the application and back? Do they see the validation errors, or perhaps just the state of the application as-is before they write something.
+- What if a user starts filling in field 3? Do they see validation errors for fields 1-3 or only 3?
+- What to do with required fields? Should they all turn red immediately or only upon save? Or only upon submit?
+
+Idea: Three states for application fields:
+- untouched (not filled in at all)
+- edited but not valid (can save, can't submit)
+- filled in and completely valid (can submit)
+
+Idea: Only show the validation errors for the fields the user actively changes in that "session". Nothing validated when the page is opened.
```

Comment:
from UX point-of-view it's not good if we validate data only on change. next time applicant comes back to the form, if they don't see the errors it would suggest that the inputs are actually valid

Reply:
This might not have been the best way to express it, what the document should say is "if we have not started typing anything in the form, then we do not show validation messages, once we do, we validate fields on change and after change", hopefully I was able to explain it well.

Reply:
Yeah, it's true that if the user comes back later, they would not initially see any errors. However you would always get them on send (with the problems due to delay)